### ISSUE TITLE
Circular references lead to "too much recursion"

### ISSUE THREAD

**Issue by s-tittel** [rdf-ext/shacl-engine#27]
Hi and thanks for your great library!

When validating a data graph that contains circular references (e.g. hasPart / isPartOf relations) , I get:
```
Uncaught (in promise) InternalError: too much recursion
```
Here is an example to reproduce:
```
import N3 from 'n3'
import { Validator } from 'shacl-engine'

const dataset = new N3.Store()
const parser = new N3.Parser()
parser.parse(
`
@prefix sh:      <http://www.w3.org/ns/shacl#> .
@prefix rdfs:    <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .

ex:Thing
    a sh:NodeShape, rdfs:Class ;
    sh:property [ 
	          sh:path ex:hasPart ;
	          sh:node ex:ThingPart
	        ] .

ex:ThingPart
    a sh:NodeShape ;
    sh:property [ 
	          sh:path ex:isPartOf ;
	          sh:node ex:Thing ;
	        ] .        

ex:thing
    a ex:Thing ;
    ex:hasPart ex:part .

ex:part
    ex:isPartOf ex:thing .
`,
async (error, quad) => {
    if(error) throw error
    if (quad)
        dataset.addQuad(quad)
    else {
        // create a validator instance for the shapes in the given dataset
        const validator = new Validator(dataset, { factory: N3.DataFactory })

        // run the validation process
        const report = await validator.validate({ dataset })
        console.log(`conforms: ${report.conforms}`)
    }
})
```

Edit: I think this directly relates to https://github.com/rdf-ext/grapoi/issues/13

**Comment by bergos** [comment]
@s-tittel  Thanks for reporting! Sorry that it took a bit longer, but I was on vacation. I had a brief look, could reproduce the problem, and already have a solution in mind. It's similar to the `grapoi` problem but a bit different. Besides the visited focus node, we also need to keep the term of the shape in memory, as multiple shapes will visit the same focus node. And `grapoi` handles only a single `sh:path` walk and doesn't have the shape content. The check needs to be implemented in the `shacl-engine`. I will have a closer look in the next few days and implement a fix.

**Comment by s-tittel** [comment]
Hi, thanks! Just fyi, the rdf-validate-shacl library had the same issue which had been fixed by [this PR](https://github.com/zazuko/rdf-validate-shacl/pull/137). I am curious about your solution since I would like to switch to your library because of its ~200 KB smaller bundle size :-)

**Comment by bergos** [comment]
I have a working solution, but I want to double-check if it is correct and works for some edge cases. A report edge case is mentioned [here](https://github.com/rdf-ext/shacl-engine/blob/61f7d8955d87b9f353795c77ed9519b60a76d06a/test/assets/data-shapes/core/validation-reports/shared.ttl#L11-L13). That was easy to solve as `grapoi` keeps track of the path and it only requires adding the path to the id of the processed shapes/focus nodes. I will get back to you once I finish my further tests.

**Comment by bergos** [comment]
I merged my changes and released a new patch version. No parameter like `maxNodeChecks` was required. The test was adapted a bit. Some names have been changed, making it clearer what is a shape and what is a class. Also, blank nodes have been changed to named nodes, which makes it easier to compare the expected results. And one `sh:node` has been replaced with a `sh:or`. A separate report is generated internal for logical shapes (or, and, ...). That use case should also be covered by the test.
